File: clients/reactive-dart/native/ios/objective-c/OptoSyncBackgroundScheduler.m
#import "OptoSyncBackgroundScheduler.h"

static NSString * const OptoSyncBackgroundChannel = @"opto-sync/background";

@interface OptoSyncBackgroundScheduler ()
@property(nonatomic, copy) NSString *identifier;
@property(nonatomic, assign) NSInteger budgetMilliseconds;
@property(nonatomic, copy) OptoSyncFlutterEngineFactory engineFactory;
@end

@implementation OptoSyncBackgroundScheduler

- (instancetype)initWithIdentifier:(NSString *)identifier
                budgetMilliseconds:(NSInteger)budgetMilliseconds
                      engineFactory:(OptoSyncFlutterEngineFactory)engineFactory {
  NSParameterAssert(identifier.length > 0);
  NSParameterAssert(budgetMilliseconds >= 1000 && budgetMilliseconds <= 600000);
  self = [super init];
  if (self) {
    _identifier = [identifier copy];
    _budgetMilliseconds = budgetMilliseconds;
    _engineFactory = [engineFactory copy];
  }
  return self;
}

- (BOOL)registerTask {
  __weak typeof(self) weakSelf = self;
  return [[BGTaskScheduler sharedScheduler]
      registerForTaskWithIdentifier:self.identifier
                         usingQueue:nil
                      launchHandler:^(__kindof BGTask *task) {
    __strong typeof(weakSelf) self = weakSelf;
    if (!self || ![task isKindOfClass:[BGProcessingTask class]]) {
      [task setTaskCompletedWithSuccess:NO];
      return;
    }
    [self handleTask:(BGProcessingTask *)task];
  }];
}

- (BOOL)scheduleAfter:(NSDate *)earliestBeginDate error:(NSError **)error {
  BGProcessingTaskRequest *request =
      [[BGProcessingTaskRequest alloc] initWithIdentifier:self.identifier];
  request.requiresNetworkConnectivity = YES;
  request.requiresExternalPower = NO;
  request.earliestBeginDate = earliestBeginDate;
  return [[BGTaskScheduler sharedScheduler] submitTaskRequest:request error:error];
}

- (void)handleTask:(BGProcessingTask *)task {
  FlutterEngine *engine = self.engineFactory();
  FlutterMethodChannel *channel = [FlutterMethodChannel
      methodChannelWithName:OptoSyncBackgroundChannel
            binaryMessenger:engine.binaryMessenger];
  __block BOOL completed = NO;
  __block BOOL runStarted = NO;

  void (^complete)(BOOL) = ^(BOOL success) {
    @synchronized (task) {
      if (completed) return;
      completed = YES;
    }
    dispatch_async(dispatch_get_main_queue(), ^{
      [channel setMethodCallHandler:nil];
      [task setTaskCompletedWithSuccess:success];
      [engine destroyContext];
    });
  };

  void (^beginRunAfterDartIsReady)(void) = ^{
    @synchronized (task) {
      if (completed || runStarted) return;
      runStarted = YES;
    }
    [channel invokeMethod:@"runOnce"
                arguments:@{ @"budgetMilliseconds": @(self.budgetMilliseconds) }
                   result:^(id _Nullable result) {
      complete(![result isKindOfClass:[FlutterError class]]);
    }];
  };

  [channel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
    if (![call.method isEqualToString:@"ready"]) {
      result(FlutterMethodNotImplemented);
      return;
    }
    result(nil);
    beginRunAfterDartIsReady();
  }];

  task.expirationHandler = ^{
    [channel invokeMethod:@"cancel" arguments:nil];
    complete(NO);
  };

  if (![engine runWithEntrypoint:@"optoSyncBackgroundMain"]) {
    complete(NO);
  }
}

@end

File: clients/reactive-dart/native/ios/swift/OptoSyncBackgroundScheduler.swift
import BackgroundTasks
import Flutter
import Foundation

/// Reference iOS integration for a bounded opto-sync HTTP push/pull cycle.
/// Register once during application launch; scheduling remains discretionary.
public final class OptoSyncBackgroundScheduler {
  public static let methodChannelName = "opto-sync/background"

  private let identifier: String
  private let engineFactory: () -> FlutterEngine
  private let budgetMilliseconds: Int

  public init(
    identifier: String,
    budgetMilliseconds: Int = 25_000,
    engineFactory: @escaping () -> FlutterEngine
  ) {
    precondition(!identifier.isEmpty)
    precondition((1_000...600_000).contains(budgetMilliseconds))
    self.identifier = identifier
    self.engineFactory = engineFactory
    self.budgetMilliseconds = budgetMilliseconds
  }

  /// Call exactly once before application launch finishes.
  @discardableResult
  public func register() -> Bool {
    BGTaskScheduler.shared.register(
      forTaskWithIdentifier: identifier,
      using: nil
    ) { [weak self] task in
      guard
        let self,
        let processingTask = task as? BGProcessingTask
      else {
        task.setTaskCompleted(success: false)
        return
      }
      self.handle(processingTask)
    }
  }

  public func schedule(earliestBeginDate: Date? = nil) throws {
    let request = BGProcessingTaskRequest(identifier: identifier)
    request.requiresNetworkConnectivity = true
    request.requiresExternalPower = false
    request.earliestBeginDate = earliestBeginDate
    try BGTaskScheduler.shared.submit(request)
  }

  private func handle(_ task: BGProcessingTask) {
    let engine = engineFactory()
    let channel = FlutterMethodChannel(
      name: Self.methodChannelName,
      binaryMessenger: engine.binaryMessenger
    )
    let stateLock = NSLock()
    var completed = false
    var runStarted = false

    func complete(_ success: Bool) {
      stateLock.lock()
      guard !completed else {
        stateLock.unlock()
        return
      }
      completed = true
      stateLock.unlock()
      DispatchQueue.main.async {
        channel.setMethodCallHandler(nil)
        task.setTaskCompleted(success: success)
        engine.destroyContext()
      }
    }

    func beginRunAfterDartIsReady() {
      stateLock.lock()
      guard !completed && !runStarted else {
        stateLock.unlock()
        return
      }
      runStarted = true
      stateLock.unlock()
      channel.invokeMethod(
        "runOnce",
        arguments: ["budgetMilliseconds": budgetMilliseconds]
      ) { result in
        complete(!(result is FlutterError))
      }
    }

    channel.setMethodCallHandler { call, result in
      guard call.method == "ready" else {
        result(FlutterMethodNotImplemented)
        return
      }
      result(nil)
      beginRunAfterDartIsReady()
    }

    task.expirationHandler = {
      channel.invokeMethod("cancel", arguments: nil)
      complete(false)
    }

    guard engine.run(withEntrypoint: "optoSyncBackgroundMain") else {
      complete(false)
      return
    }
  }
}
